File: PartitionEqualSubsetSum/partition_equal_subset_sum.py
#!/usr/bin/env python3

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        nums_sum = sum(nums)
        if nums_sum %2 != 0:
            return False

        target = nums_sum//2
        table = [False]*(target+1)
        table[0] = True

        for num in nums:
            for i in range(target, num-1, -1):
                table[i] = table[i] or table[i-num]
        
        return table[target]

File: PartitionEqualSubsetSum/partition_equal_subset_sum.cpp
#include <numeric>
#include <vector>
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum{std::accumulate(nums.begin(), nums.end(), 0)};
        if (sum & 1 != 0)
            return false;
        int target{sum/2};
        std::vector<bool> dp(target+1, false);
        dp[0] = true;
        for (int num : nums){
            for (int i = target; i > 0; --i){
                if (num <= i){
                    dp[i] = dp[i] || dp[i-num];
                }
            }
        }
        return dp[target];
    }
};

